service/service.go

```diff
@@ -98,16 +99,43 @@ func New(tarball io.Reader, env map[string]string, options ...Option) (*Service,
 	if err := s.setOptions(options...); err != nil {
 		return nil, err
 	}
+	// untar tarball to retrieve mesg.yml
 	if err := s.saveContext(tarball); err != nil {
 		return nil, err
 	}
 	defer os.RemoveAll(s.tempPath)
-
 	def, err := importer.From(s.tempPath)
 	if err != nil {
 		return nil, err
 	}
 
+	// XXX: remove .git folder from repo.
+	// It makes docker build iamge id same between repo clones.
+	if err := os.RemoveAll(filepath.Join(s.tempPath, ".git")); err != nil {
+		return nil, err
+	}
+
+	// XXX: change the access and modification times of service to get constant hash
+	zerot := time.Time{}
+	if err := filepath.Walk(s.tempPath, func(path string, _ os.FileInfo, _ error) error {
+		return os.Chtimes(path, zerot, zerot)
+	}); err != nil {
+		return nil, err
+	}
+
+	// tar it back without any compression to get constant hash
+	// despite used compression from tarball
+	t, err := archive.Tar(s.tempPath, archive.Gzip)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Let's move this code from line 112 (remove .git) to line 131 (the creation of the tar) to `api/deploy.go`.
`api/deploy.go` should download AND clean the downloaded stuff (like it's already doing for git), and create a clean tarball to `Service.New`. In the future, we can pass a path between `api/deploy.go` and `Service.New` to avoir tar/untar again, but this should be done in a separate PR.

Reply:
Also, the check for one folder in the root folder should also be moved to `api/deploy`:
https://github.com/mesg-foundation/core/blob/6700ccf4767899795308a5572d64bd57a0b45ff8/service/service.go#L220-L222

Reply:
In the future for optimization reason, `Service.New` should accept a path instead of an `io.reader` to avoid unnecessary untar/tar.
This way, `api/deploy` downloads from wherever to a local folder, clean it, pass it to `Service.New`.
`Service.New` can create a tar if needed to calculate hash.